src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -126,5 +147,112 @@ public static partial class JsonSerializer
+            {
+                // Check we are not parsing into immutable.
+                if (state.Current.TempDictionaryValues != null)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(state.Current.JsonPropertyInfo!.DeclaredPropertyType);
+                }
+
+                if (state.Current.KeyName != null || isPreserved)
+                {
+                    state.Current.KeyName = null;
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                }
+
+                MarkAsPreserved(ref state.Current);
+                state.Current.ReadMetadataValue = true;
+            }
+            else if (metadata == MetadataPropertyName.Ref)
+            {
+                if (state.Current.KeyName != null || isPreserved)
+                {
+                    state.Current.KeyName = null;
+                    ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties();
+                }
+
+                state.Current.ShouldHandleReference = true;
+                state.Current.ReadMetadataValue = true;
+            }
+        }
+
+        private static void ResolveMetadataOnObject(ReadOnlySpan<byte> propertyName, MetadataPropertyName meta, ref ReadStack state, ref Utf8JsonReader reader, JsonSerializerOptions options)
+        {
+            if (meta == MetadataPropertyName.NoMetadata)
+            {
+                if (state.Current.IsPreservedArray)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(state.Current.JsonClassInfo!.PropertyCache!["Values"].DeclaredPropertyType);
+                }
+                // Regular property, call main logic for HandlePropertyName.
+                HandlePropertyNameDefault(propertyName, ref state, ref reader, options);
+            }
+            else if (meta == MetadataPropertyName.Id)
+            {
+                if (state.Current.PropertyIndex > 0 || state.Current.IsPreserved)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                }
+
+                JsonPropertyInfo info = JsonPropertyInfo.s_metadataProperty;
+                info.JsonPropertyName = propertyName.ToArray();
+                state.Current.JsonPropertyInfo = info;
+
+                state.Current.ReadMetadataValue = true;
+                MarkAsPreserved(ref state.Current);
+            }
+            else if (meta == MetadataPropertyName.Values)
+            {
+                // Preserved JSON arrays are wrapped into JsonPreservedReference<T> where T is the original type of the enumerable
+                // and Values is the actual enumerable instance being preserved.
+                JsonPropertyInfo info = state.Current.JsonClassInfo!.PropertyCache!["Values"];
+                info.JsonPropertyName = propertyName.ToArray();
```

Comment:
Anything pulled from this cache can't be changed since it is shared across every serialization (e.g. multiple threads may access this)

Reply:
it is ok for this property that the name remains the same across serializations.

Reply:
A property value changes via `info.JsonPropertyName = propertyName.ToArray()`. In a multi-threaded case they would collide.

Reply:
I assume the JsonPropertyInfo instance here can be used for more than one array property -- each with a different name -- so the JsonPropertyInfo instance here needs to be only for that array (otherwise there will be naming collisions). Will the instance be correct\unique, or is there an issue here?

Reply:
Yes and no, it is used for each array property with preserve semantics **after** it was wrapped into `JsonPreservedReference<T>`  where `T` is the original array's property type.
For preserved semantics we always call the property `$values` when processing `JsonPreservedReference<T>.Values` since that is what it represents in `"arrayProperty": { "$id": "1", "$values": [...] }`.

Reply:
Essentially it emulates [the behavior in our code](https://github.com/dotnet/runtime/pull/655/files/c965e2cb60093c32139ef6543b8be6c6068a87df#diff-90151f731bdfb61249742dc1644ba8eeR96) to support writing the property name in the JSON path for regular properties. 
I might as well would copy the condition on line 104 here and replace propertyName.ToArray() for a constant to avoid allocation at runtime.

Reply:
If the property name is always "$values" then the code is fine, although using a constant as you mention would be an improvement.

Reply:
We may be able to use a cached JsonPropertyInfo as well - to avoid any memory copy - but we can do that later after the initial code is in.